compiler/rustc_target/src/spec/mod.rs

```diff
@@ -2185,21 +2197,34 @@ fn add_link_args_iter(
     match flavor {
         LinkerFlavor::Gnu(cc, lld) => {
             assert_eq!(lld, Lld::No);
+            assert_matches!(cc, Cc::No | Cc::Yes);
             insert(LinkerFlavor::Gnu(cc, Lld::Yes));
+            if cc == Cc::Yes {
+                insert(LinkerFlavor::Gnu(Cc::Clang, Lld::No));
+                insert(LinkerFlavor::Gnu(Cc::Clang, Lld::Yes));
+            }
         }
         LinkerFlavor::Darwin(cc, lld) => {
             assert_eq!(lld, Lld::No);
+            assert_matches!(cc, Cc::No | Cc::Yes);
             insert(LinkerFlavor::Darwin(cc, Lld::Yes));
+            if cc == Cc::Yes {
+                insert(LinkerFlavor::Darwin(Cc::Clang, Lld::No));
+                insert(LinkerFlavor::Darwin(Cc::Clang, Lld::Yes));
+            }
         }
         LinkerFlavor::Msvc(lld) => {
             assert_eq!(lld, Lld::No);
             insert(LinkerFlavor::Msvc(Lld::Yes));
         }
-        LinkerFlavor::WasmLld(..)
-        | LinkerFlavor::Unix(..)
-        | LinkerFlavor::EmCc
-        | LinkerFlavor::Bpf
-        | LinkerFlavor::Ptx => {}
+        LinkerFlavor::WasmLld(cc) | LinkerFlavor::Unix(cc) => {
+            assert_matches!(cc, Cc::No | Cc::Yes);
+            if cc == Cc::Yes {
+                insert(LinkerFlavor::Gnu(Cc::Clang, Lld::No));
+                insert(LinkerFlavor::Gnu(Cc::Clang, Lld::Yes));
```

Comment:
Possibly still WIP: these linker args should have a `WasmLld` or `Unix` principal flavor here, rather than `Gnu`?